### PR TITLE
LoadBalancing of GRPC demo

### PR DESCRIPTION
**Pull request by karol-kokoszka**
+ reflection support in grpc-book-server
+ bug fix in grpc-book-server
+ Dockerfile for grpc-book server
+ Dockerfile for book-cli
+ K8S deployment for server and cli
+ README describing minikube and linkerd isntallation

### CODE REVIEW

File: _k8s/deployment/grpc-book-cli.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: grpc-book-cli
+  template:
+    metadata:
+      labels:
+        app: grpc-book-cli
+    spec:
```
Comment (rjeczalik) [inline _k8s/deployment/grpc-book-cli.yaml:16]: Can we have a helm chart instead? Manifests cannot be parametrized.

Also you need ingress resource definition, ingress is a requirement for api services.

File: _k8s/docker/grpc-book-srv.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FROM golang:1.17-alpine
+
+ENV PATH="/src/sc-poc/bin/:${PATH}"
+
+ADD . /src/sc-poc/
+WORKDIR /src/sc-poc/
+
+RUN apk add --update build-base curl bash
```
Comment (rjeczalik) [inline _k8s/docker/grpc-book-srv.Dockerfile:8]: please add `--no-cache` for `apk add`, so the archvies won't get saved into the image

**Review by rjeczalik — CHANGES_REQUESTED** [review]
I think the most important part of k8s orchestration is the usage of liveness and readiness probes.

And those probes should be actually implemented by the service you build, not a dummy handler.

So I think we need to also add probes to k8s resources and code handlers for them

**Comment by karol-kokoszka** [comment]
I changed the client a bit, so that it executes 10 rpc calls in one execution to better show load balancing.

Output from the setup with just a service in front of services:
```
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 5 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 5 a t
2022/02/21 09:17:02 6 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 5 a t
2022/02/21 09:17:02 6 a t
2022/02/21 09:17:02 7 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 5 a t
2022/02/21 09:17:02 6 a t
2022/02/21 09:17:02 7 a t
2022/02/21 09:17:02 8 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 5 a t
2022/02/21 09:17:02 6 a t
2022/02/21 09:17:02 7 a t
2022/02/21 09:17:02 8 a t
2022/02/21 09:17:02 9 a t
2022/02/21 09:17:02 1 a t
2022/02/21 09:17:02 2 a t
2022/02/21 09:17:02 3 a t
2022/02/21 09:17:02 4 a t
2022/02/21 09:17:02 5 a t
2022/02/21 09:17:02 6 a t
2022/02/21 09:17:02 7 a t
2022/02/21 09:17:02 8 a t
2022/02/21 09:17:02 9 a t
2022/02/21 09:17:02 10 a t
```
^^^ It just shows that all the request calls goes to a single POD (as the connection is established with the service and it connects this particular CLI with a single POD)

**Comment by karol-kokoszka** [comment]
... and the output from the deployment with Linkerd sidecars applied:
```
bash-5.1# ./book-cli.dev --addr grpc-book-svc:50051
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 2 a t
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 2 a t
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 2 a t
2022/02/21 09:26:17 3 a t
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 1 a t
2022/02/21 09:26:17 2 a t
2022/02/21 09:26:17 3 a t
```

Traffic is distributed between different pods.
Connection is established with the service, but requests are routed to different pods.

File: _k8s/deployment/grpc-book-cli.yaml (hunk 1 above)
Comment (karol-kokoszka) [reply _k8s/deployment/grpc-book-cli.yaml:16]: Hey @rjeczalik.
Of course, we can add helm support. But I'm not sure if it should be added in this PR.
The motivation I had with this PR is to build minimum k8s deployment that let me to test GRPC load balancing.

I want to test/show it two ways:
- with Linkerd control plane installed on the cluster + sidecar proxies on GRPC server
- and just a k8s service where the connection is created to the service (to show that all the calls will reach just a single POD)

Expectation I have is that the first approach will distribute the calls to different backends.
The second approach will always reach just a single POD.

I can add helms, but I'm not sure if this is a good moment to polish the k8s deployment.

File: _k8s/docker/grpc-book-srv.Dockerfile (hunk 2 above)
Comment (karol-kokoszka) [reply _k8s/docker/grpc-book-srv.Dockerfile:8]: ofc, added

**Comment by karol-kokoszka** [comment]
> I think the most important part of k8s orchestration is the usage of liveness and readiness probes.

There is no doubt that liveness and readiness probes must be created on the deployment when we have the service, but it was not the goal of the PR to build the complete prod ready deployment.
To avoid misunderstood I'll change the PR name.

> And those probes should be actually implemented by the service you build, not a dummy handler.

> So I think we need to also add probes to k8s resources and code handlers for them

**Review by rjeczalik — COMMENTED** [review]
removing req change

**Review by rjeczalik — APPROVED** [review]

**Comment by mmatczuk** [comment]
let's move that under grpc/k8s

**Comment by mmatczuk** [comment]
@karol-kokoszka

**Comment by karol-kokoszka** [comment]
@mmatczuk Done

**Pull request merged**